### PR TITLE
connection pool improvements

### PR DESCRIPTION
**Pull request by wgallagher**

### CODE REVIEW

**Comment by wgallagher** [comment]
```
  - allow a minimum connection pool size to be specified
  - round-robin between pool connections
  - kill connections after a specified number of uses
```

**Comment by wgallagher** [comment]
https://lyftme.atlassian.net/browse/CORESRV-367

File: server.go [hunk 1]
```diff
@@ -114,10 +116,10 @@ func (server *mongoServer) AcquireSocket(poolLimit int, timeout time.Duration) (
...
 			return nil, false, errPoolLimit
 		}
-		if n > 0 {
-			socket = server.unusedSockets[n-1]
-			server.unusedSockets[n-1] = nil // Help GC.
-			server.unusedSockets = server.unusedSockets[:n-1]
+		if n > 0 && len(server.liveSockets) >= minPoolLimit {
+			socket = server.unusedSockets[0]
+			server.unusedSockets[0] = nil // Help GC.
+			server.unusedSockets = server.unusedSockets[1:]
```
Comment (mattklein123) [inline server.go]: I'm not exactly sure how this is implemented, but presumably this is going to cause a full shift of all elements. We might consider switching this from a slice to a linked list. Not a huge deal but I thought I would mention it.

File: socket.go [hunk 2]
```diff
@@ -264,10 +266,17 @@ func (socket *mongoSocket) Release() {
...
+			socket.maxUses -= 1
+			if socket.maxUses == 0 {
+				recycle = false
+			}
+		}
 		socket.Unlock()
 		socket.LogoutAll()
 		// If the socket is dead server is nil.
-		if server != nil {
+		if server != nil && recycle {
```
Comment (mattklein123) [inline socket.go:279]: I think we should close() the socket here if server != nil and we don't recycle it?

File: socket.go [hunk 3]
```diff
@@ -278,6 +278,8 @@ func (socket *mongoSocket) Release() {
 		// If the socket is dead server is nil.
 		if server != nil && recycle {
 			server.RecycleSocket(socket)
+		} else {
+			socket.Close()
```
Comment (mattklein123) [inline socket.go:282]: this is safe if socket is already closed?

**Comment by mattklein123** [comment]
👍

File: socket.go (hunk 3 above)
Comment (wgallagher) [inline socket.go:282]: it looks like it just logs if a socket is closed twice

**Pull request closed without merging**